### ISSUE TITLE
Two questions about training process

### ISSUE THREAD

**Issue by cdyrhjohn** [nagadomi/waifu2x#108]
1. Your model was trained on a 6000-image dataset. So for each image, how did you crop them?
2. What's your training/validation accuracy after you finished training the model?

**Comment by nagadomi** [comment]
1. No crop. (but training code randomly crops image from original size to 256x256 subimages -> 46x46 patches in each iteration)
2. I don't have logs. Bechmark results can be found at [Benchmark results](https://github.com/nagadomi/waifu2x/blob/df4887d719a7645bc9a440cbd022b4ca50052df4/appendix/benchmark.md). Unfortunately that benchmark dataset is closed data.

**Comment by cdyrhjohn** [comment]
Thanks for your answer. In fact, I'm a beginner of Deep learning and struggling to finish my final project of machine learning course. I tried to follow the instruction of the paper of Chao Dong using 3 conv layers. How should I tune my hyperparameter? I have tried to use the parameter given by Dong, but some of the predicted images had noisy points. Could you give me some advise? Thank you. For my training set, I used 1000 256*256 images from ImageNet and croped each image as 64 32*32images(not randomly).

**Comment by nagadomi** [comment]
I think SRCNN based neural networks is hard to optimize. I used Adam([arxiv](http://arxiv.org/abs/1412.6980), [Torch implementatin](https://github.com/torch/optim/blob/master/adam.lua)) to optimize waifu2x's model. Adam is easy to converge this task than mometum-SGD, in my experience.

**Comment by cdyrhjohn** [comment]
Yep, I've read part your code and I also tried the Adam for my gradient descent. But Adam is recommended to use its default learning rate, so I am not sure whether I should change it a little bit for SRCNN. Another question is about your LeakyReLU activation, does that dramatically improve your performance than ReLU? Finally, I want to ask for an efficient way of tuning the parameter(using a small number of training data or maybe using a small number of epoch?) Thanks a lot.

**Comment by nagadomi** [comment]
LeakyReLU does not dramatically improve performance in this small network(3 layers or 7 layers).
I used small dataset for debug(and little tuning), and full dataset for tuning hyperprameters.

I think the most important things in this task is to make pairwise teacher data correctly.
![4_x](https://cloud.githubusercontent.com/assets/287255/15071275/47f0f944-13c5-11e6-8b60-df6b3b0e2337.png)
![4_y](https://cloud.githubusercontent.com/assets/287255/15071277/49266bf0-13c5-11e6-8347-eb6616581616.png)
 If training code has bug in generating pairwise data, trained model will generate low quaily images.

**Comment by cdyrhjohn** [comment]
Anyway, thank u a lot. So do I need to keep the learning rate of Adam as default 0.001? Or tune it a little bit?

**Comment by nagadomi** [comment]
I used low learning rate 0.0005 ~ 0.00001 (with learning rate decay).

**Comment by cdyrhjohn** [comment]
How should I set my batch size? You said that it helps if we set it 2~4. How about making it a larger number?  The efficiency was really low when I used size 4.

**Comment by nagadomi** [comment]
current my settings:
input: 3x46x46
output: 3x32x32 (convolution layer does not have padding, so output size is smaller than input size.)
optimizer: Adam
batch_size: 8
learning_rate: 0.0005

**Comment by cdyrhjohn** [comment]
Could you please explain why Waifu2x has excellent performance on "anime pictures" (The lines in anime pictures are somehow smoothed?)? I've finished my simplified SRCNN model and I find my 3-layer-net even have better performance on "realistic photos".

**Comment by cdyrhjohn** [comment]
What's the different when you deal with anime pictures and realistic photos respectively?

**Comment by nagadomi** [comment]
That is learned by the neural network.
waifu2x has the photo model(`models/photo`). It was trained by the same code as anime style art.
(http://waifu2x.udp.jp/ Style=Photo)

Did you use gaussian blur to generate dataset?
I think that the original SRCNN uses bicubic interplation. 
See SRNN_train.zip(./generate_train.m) http://mmlab.ie.cuhk.edu.hk/projects/SRCNN.html

**Comment by leilei-** [comment]
i'm also curious on the training when using nearest neighbor interpolation to upscale pixel art.  There's this "MagicPony" neural network image improver out there getting some hype lately for doing just that at attempting to make realistic interpretaions of pixel art or pixelized art

**Comment by nagadomi** [comment]
I guess MagickPony is something like Deep Dream or [Perceptual Loss](http://arxiv.org/abs/1603.08155). It is able to generate high resolution textures from (memorized) training images.
and, I am thinking about developing a inverse dithering(color reduction) filter but that is a very low priority task for me at the moment.

**Comment by cdyrhjohn** [comment]
SRCNN is also used for denoising? Or you did something else?

**Comment by nagadomi** [comment]
The model of waifu2x is not the same as SRCNN. It's VGG style, 7 layers CNN. waifu2x uses that model for denoising and upscaling. And I think that deblurring task is more difficult than denoising task.

**Comment by ProGamerGov** [comment]
@nagadomi What was the resolution of the images used in your training data? How does the resolution of the training data affect training an art model, and what is the best resolution for training data? Do I need to find images with sizes like 2k, or 4k for my data set? Or does Waifu2x not require such a high image size for successful training?

**Comment by cdyrhjohn** [comment]
@ProGamerGov Nope, SRCNN only learns how to increase the resolution, like a filter. Once you have the filter, you can increase the resolution of any image. Once you know how to drive, you'll never mind you drive BMW or Benz or Landrover.

**Comment by ProGamerGov** [comment]
@cdyrhjohn , so is there a limit on the size/resolution of the training data I can use? Will higher resolutions provide better results?

**Comment by cdyrhjohn** [comment]
@ProGamerGov no limit; no better result, but much more training time; waifu2x use 46x46 resolution dataset.
